### PR TITLE
[SPARK-26509][SQL] Parquet DELTA_BYTE_ARRAY is not supported in Spark 2.x's Vectorized Reader

### PR DESCRIPTION
**Pull request by nandorKollar**
## What changes were proposed in this pull request?

Implement Parquet delta encoding for vectorized interface, which is needed for V2 pages. The implementation simply delegates the decoding to the Parquet implementation.

## How was this patch tested?

Added new test case for delta encoding, ran unit tests

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @rdblue

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  public void readDoubles(int total, WritableColumnVector c, int rowId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void readBinary(int total, WritableColumnVector c, int rowId) {
+    for (int i = 0; i < total; i++) {
+      Binary binary = deltaByteArrayReader.readBytes();
+      ByteBuffer buffer = binary.toByteBuffer();
+      if (buffer.hasArray()) {
```
Comment (attilapiros) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java]: I am a bit uncertain here but I have tried `binary.getBytes()` and it worked. 
I know currently the buffer is backed by a byte array. 

So my questions:
- Can we use `binary.getBytes()` for both cases?
- What would be its disadvantages?

**Comment by HyukjinKwon** [comment]
ping @nandorKollar

**Comment by kiszk** [comment]
ping @nandorKollar

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala [hunk 2]
```diff
@@ -114,4 +118,36 @@ class ParquetEncodingSuite extends ParquetCompatibilityTest with SharedSQLContex
...
+        spark.createDataFrame(data).write.options(extraOptions).mode("overwrite").parquet(path)
+
+        val blockMetadata = readFooter(new Path(path), hadoopConf).getBlocks.asScala.head
+        val columnChunkMetadataList = blockMetadata.getColumns.asScala
+
+        // Verify that indeed delta encoding is used for each column
+        assert(columnChunkMetadataList.length === 3)
+        assert(columnChunkMetadataList(0).getEncodings.contains(Encoding.DELTA_BINARY_PACKED))
+        assert(columnChunkMetadataList(1).getEncodings.contains(Encoding.DELTA_BINARY_PACKED))
+        assert(columnChunkMetadataList(2).getEncodings.contains(Encoding.DELTA_BYTE_ARRAY))
```
[the remark this answers is not in the document]
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala]: Done in https://github.com/apache/spark/pull/27316/commits/947c6f732ea127d6f60f0371138127bda6e78770

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java (hunk 1 above)
Comment (xuanyuanking) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java]: Find the discussion here: https://github.com/apache/spark/pull/21070#discussion_r181882476.

**Pull request closed without merging**